kaggle_environments/main.py

```diff
@@ -113,45 +122,75 @@ def action_act(args):
 
     # Pass empty steps in here because we just need the configuration from the environment
     env = make(args.environment, args.configuration, [], args.debug)
-    config = env.configuration
-    timeout = config.actTimeout
 
     if cached_agent is None or cached_agent.raw != raw:
         cached_agent = Agent(raw, env)
-        timeout = config.agentTimeout
     observation = utils.get(args.state, dict, {}, ["observation"])
-    action = cached_agent.act(observation, timeout)
+    action, log = cached_agent.act(observation)
     if isinstance(action, errors.DeadlineExceeded):
         action = "DeadlineExceeded"
     elif isinstance(action, BaseException):
         action = "BaseException::" + str(action)
 
+    if args.log_path is not None:
+        with open(args.log_path, mode="a") as log_file:
+            json.dump([log], log_file)
+            log_file.write(",\n ")
+
     return {"action": action}
 
 
 def action_step(args):
-    env = make(args.environment, args.configuration, args.steps, args.debug)
+    env = make(args.environment, args.configuration, args.steps, args.logs, args.debug)
     runner = env.__agent_runner(args.agents)
     env.step(runner.act())
+    if args.log_path is not None:
+        with open(args.log_path, mode="a") as log_file:
+            json.dump(env.logs[-1], log_file)
+            log_file.write(",")
     return render(args, env)
 
 
 def action_run(args):
-    env = make(args.environment, args.configuration, args.steps, args.debug)
+    env = make(args.environment, args.configuration, args.steps, args.logs, args.debug)
     env.run(args.agents)
+    if args.log_path is not None:
+        with open(args.log_path, mode="w") as log_file:
+            json.dump(env.logs, log_file)
     return render(args, env)
 
 
 def action_load(args):
+    if args.log_path is not None:
+        with open(args.log_path, mode="r") as log_file:
+            args.logs = json.load(log_file)
+
     if args.in_path is not None:
         with open(args.in_path, mode="r") as replay_file:
             json_args = json.load(replay_file)
-        env = make(json_args["name"], json_args["configuration"], json_args["steps"], args.debug)
+        env = make(json_args["name"], json_args["configuration"], json_args["steps"], args.logs, args.debug)
     else:
-        env = make(args.environment, args.configuration, args.steps, args.debug)
+        env = make(args.environment, args.configuration, args.steps, args.logs, args.debug)
     return render(args, env)
 
 
+disposed = True
+
+
+# This method is only called at the end of an episode to write the final array brace in the logs file and dispose the cached agent to force reinitialization
+def action_dispose(args):
+    global cached_agent, disposed
+    if disposed:
+        return "Already disposed"
+
+    cached_agent = None
+    if args.log_path is not None:
+        with open(args.log_path, mode="a") as log_file:
+            log_file.write("]")
```

Comment:
nit: What's the advantage to iteratively opening / writing to the log file vs just storing everything and writing it all at the end?

Reply:
Doing it all at the end is probably better and easier to follow, I'll give that a shot.

Reply:
Looking into this more there are a few issues with writing everything at the end. The dispose method doesn't have access to the environment, and the agent image never gets the full environment so it must track its own logs. We could store the logs in memory rather than on disk, but this solution has the same problems with global state and also means that we wouldn't get logs if the container crashes, where the iterative file approach will output malformed but still perfectly usable json.

I would like to add an execution context object for tracking this sort of stuff in the future that has information about the current episode run, so I'll probably fold in a refactor on step and act logs into that project when we get to it.